### ISSUE TITLE
Async App mysteriously started failing WS Handshake on June 16th, 1 AM EST

### ISSUE THREAD

**Issue by Zolnoor** [slackapi/bolt-python#935]
Hello all.

I have a mysterious issue that I've been poking and prodding at for a month now with no help. I even opened a Slack support ticket but didn't get much help besides "maybe it's your openssl?" Which, I agree, it might be! But I'm not sure how.

Around June 16th, 1 am, my async websocket Slack app was disconnected from Slack servers and unable to reconnect. It had been successfully deployed with no changes to the source code or infrastructure supporting it for a week prior to this. Every attempt to redeploy this application, from rolling the source code back to several months ago to trying different docker images all results in the same thing: It fails with an error message like this:

`aiohttp.client_exceptions.WSServerHandshakeError: 404, message='Invalid response status', url=URL('wss://wss-primary.slack.com/link/?ticket=431d39cc-275e-4cbd-87d8-49aa12d7455e&app_id=a6b511b5487b4fa3cf80b88c00b1275fdfa279ceeae89c7f941cc33a53a10ad6')`

I run my apps in two environments: Non-prod and prod. Our non-prod environment works fine, even with the same app credentials. We then promote the exact same AMI and Docker Image to production where it will fail. I work at a rather large company and I've worked with both my networking and proxy teams trying to identify an issue on our side, because obviously everything I've said points to an issue in our environments. We are unable to find any discernible difference!

I do know that our production environment can reach Slack. Because the `apps.connections.open` call works and we are returned that wss-primary ticket url! However it's the handshake that follows that fails.

### Reproducible in:

slack-bolt==1.18.0
slack-sdk==3.21.3
slackeventsapi==3.0.1
aiohttp==3.8.4

Python 3.9.5

1 SMP Thu May 18 11:23:40 UTC 2023

OpenSSL 1.1.1f  31 Mar 2020

### Expected result:

Deploying the same app in both environments should work.

### Actual result:

Deploying only works in one environment despite them being identical (From what we can tell).

```
DEBUG:slack_bolt.AsyncApp:Received the following response - status: 200, headers: {'Date': 'Tue, 18 Jul 2023 14:58:56 GMT', 'Server': 'Apache', 'Vary': 'Accept-Encoding', 'x-slack-req-id': 'd52ca95edc6737540ed389dfc2e8554a', 'x-content-type-options': 'nosniff', 'x-xss-protection': '0', 'x-robots-tag': 'noindex,nofollow', 'Pragma': 'no-cache', 'Cache-Control': 'private, no-cache, no-store, must-revalidate', 'Expires': 'Sat, 26 Jul 1997 05:00:00 GMT', 'Content-Type': 'application/json; charset=utf-8', 'x-accepted-oauth-scopes': 'connections:write', 'x-oauth-scopes': 'connections:write,authorizations:read', 'Access-Control-Expose-Headers': 'x-slack-req-id, retry-after', 'Access-Control-Allow-Headers': 'slack-route, x-slack-version-ts, x-b3-traceid, x-b3-spanid, x-b3-parentspanid, x-b3-sampled, x-b3-flags', 'strict-transport-security': 'max-age=31536000; includeSubDomains; preload', 'x-slack-unique-id': 'ZLaosCk_82XGVxpsMlkZegAAEAE', 'x-slack-backend': 'r', 'referrer-policy': 'no-referrer', 'Access-Control-Allow-Origin': '*', 'Via': '1.1 slack-prod.tinyspeck.com, envoy-www-iad-wvgiyrcb, envoy-edge-iad-mmmziasf', 'Content-Encoding': 'gzip', 'Content-Length': '164', 'x-envoy-upstream-service-time': '19', 'x-backend': 'main_normal main_canary_with_overflow main_control_with_overflow', 'x-server': 'slack-www-hhvm-main-iad-idwc', 'x-slack-shared-secret-outcome': 'no-match', 'x-edge-backend': 'envoy-www', 'x-slack-edge-shared-secret-outcome': 'no-match'}, body: {'ok': True, 'url': 'wss://wss-primary.slack.com/link/?ticket=ffe945ac-1b09-4c9b-9859-92eda9a7acc8&app_id=a6b511b5487b4fa3cf80b88c00b1275fdfa279ceeae89c7f941cc33a53a10ad6'} 
```
^ 200 received in dev on July 17th followed by a connection being successfully established

```
DEBUG:slack_bolt.AsyncApp:Received the following response - status: 200, headers: {'Date': 'Tue, 18 Jul 2023 21:06:26 GMT', 'Server': 'Apache', 'x-slack-req-id': '5a4dd36d029f7a76947244f024d2021d', 'x-content-type-options': 'nosniff', 'x-xss-protection': '0', 'x-robots-tag': 'noindex,nofollow', 'Pragma': 'no-cache', 'Cache-Control': 'private, no-cache, no-store, must-revalidate', 'Expires': 'Sat, 26 Jul 1997 05:00:00 GMT', 'Content-Type': 'application/json; charset=utf-8', 'x-accepted-oauth-scopes': 'connections:write', 'x-oauth-scopes': 'connections:write,authorizations:read', 'Access-Control-Expose-Headers': 'x-slack-req-id, retry-after', 'Access-Control-Allow-Headers': 'slack-route, x-slack-version-ts, x-b3-traceid, x-b3-spanid, x-b3-parentspanid, x-b3-sampled, x-b3-flags', 'strict-transport-security': 'max-age=31536000; includeSubDomains; preload', 'x-slack-unique-id': 'ZLb-0js8Haep8OyewYUOVQAAECE', 'x-slack-backend': 'r', 'referrer-policy': 'no-referrer', 'Access-Control-Allow-Origin': '*', 'Via': '1.1 slack-prod.tinyspeck.com, envoy-www-iad-lnjvvacb, envoy-edge-iad-xvjhzhqo', 'Content-Encoding': 'gzip', 'x-envoy-upstream-service-time': '13', 'x-backend': 'main_normal main_canary_with_overflow main_control_with_overflow', 'x-server': 'slack-www-hhvm-main-iad-lcia', 'x-slack-shared-secret-outcome': 'no-match', 'x-edge-backend': 'envoy-www', 'x-slack-edge-shared-secret-outcome': 'no-match', 'Content-Length': '164', 'Connection': 'keep-alive'}, body: {'ok': True, 'url': 'wss://wss-primary.slack.com/link/?ticket=431d39cc-275e-4cbd-87d8-49aa12d7455e&app_id=a6b511b5487b4fa3cf80b88c00b1275fdfa279ceeae89c7f941cc33a53a10ad6'}  
```

```
aiohttp.client_exceptions.WSServerHandshakeError: 404, message='Invalid response status', url=URL('wss://wss-primary.slack.com/link/?ticket=431d39cc-275e-4cbd-87d8-49aa12d7455e&app_id=a6b511b5487b4fa3cf80b88c00b1275fdfa279ceeae89c7f941cc33a53a10ad6')
```

^ Exact same app creds failing the handshake in prod moments later. Also I should note it doesn't matter what app creds we use, they all fail.

Any insight would be appreciated. Even suggestions where to poke. Pulling my hair out at this one lol  😭

**Comment by seratch** [comment]
Hi @Zolnoor, thanks for your question.

As our support team mentioned, I also suspect that something related to your WebSocket connection part might be incompatible with Slack's policies. The most common issue is an old TLS version, or other SSL-related factors. However, I cannot guess more specific causes for your case.

It's indeed unusual that the same Docker image behaves differently in your production environment. However, I believe there could be factors specific to your production environment causing this issue, such as request header modifications by proxy servers.

You highlighted the difference between apps.connections.open API calls and WebSocket connection establishments. The less user-friendly 'Invalid response status' message accompanied by a 404 status is produced by the frontend layer of the real-time API server (i.e., the backend service that handles WebSocket connections). There seems to be something incompatible with the WebSocket handshake specifically, while the web API calls are allowed under different policies implemented by the frontend.

Unfortunately, I can't make further guesses. I hope you manage to identify the cause soon.

**Comment by Zolnoor** [comment]
Hi @seratch!

I really appreciate you taking the time to read and respond to this issue. We eventually swapped subnets within our VPC and we were able to connect once again. Sorry for troubling you with the non-issue, but I work in a very large corp where I have near 0 control over some of the aspects such as AWS networking resources. So, after being told by the team that controls them nothing changed (even though obviously something did), I had to exhaust all my options including coming here.

Again, appreciate your time. I'll close this issue
